docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,200 @@
+
+* ``await`` **is what will ultimately make a method asynchronous.**
+
+Failing to apply the ``await`` operator to a task will result in the async method running synchronously!  Application of ``await`` is what suspends the async method, giving back control to the method which called it.  Pay attention to compiler warnings about this.
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  That's the only reason they were allowed in the first place!  Async programming uses the ``Task`` and ``Task<T>`` objects, which provide flexibility in coordinating asynchronous work.  Throwing that out of the window with ``async void`` doesn't follow the model very well.  Here's some specific issues:
+
+    (a) Exceptions thrown in an ``async void`` method can't be caught outside of that method.
+	
+    (b) ``async void`` methods are very difficult to test.
+	
+    (c) ``async void`` methods can cause bad side effects if the caller isn't expecting them to be async.
+	
+That said, ``async void`` is perfect for event handlers where the event involves any blocking task(s).
+
+* **Tread carefully when using async lambdas in LINQ expressions**
+
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to.  The introduction of blocking tasks into this can easily result in a deadlock if not written correctly.  The nesting of asyncronous code like this can also make it more difficult to reason about the execution of the code.  Async and LINQ are powerful, but if they are abused they can make things difficult.  Clarity is always better than cleverness.
+
+* **Try to write code that "naturally" awaits blocking results**
+
+Although it is certainly easier to call other async code from an async method, not doing so correctly can result in deadlocks, blocked context threads, and significantly more complex error-handling.  The following table should provide some guidance in how to deal with blocking code from an async context:
+
+====================== ================================= =======================
+Use this...            Instead of this...                When wishing to do this
+====================== ================================= =======================
+``await``              ``Task.Wait`` or ``Task.Result``  Retreiving the result of a background task
+``await Task.WhenAny`` ``Task.WaitAny``                  Waiting for any task to complete
+``await Task.WhenAll`` ``Task.WaitAll``                  Retreiving the results of multiple tasks
+``await Task.Delay``   ``Thread.Sleep``                  Waiting for a period of time
+====================== ================================= =======================
+
+The good news is that calling async code which has no return value has no caveats.  ``JustDoesAJobAsync()`` does not need to be coordinated unless the calling method depends on its execution.  However, this would warrant a refactor, bringing up the final point...
```

Comment:
I thought that this still style of method call still needed to be assigned to a variable. No?

Why would it need refactoring?